src/julia_gc.cc

```diff
@@ -210,204 +126,471 @@ static inline void * align_ptr(void * p)
+                node->item = (*succ)->item;
+                remove_node(*succ);
+                return;
             }
             else {
-                treap_rot_left(treap);
-                treap = &L(*treap);
+                node = node->left;
             }
         }
-    }
-}
-
-static int treap_delete(treap_t ** treap, void * addr)
-{
-    while (*treap != NULL) {
-        int c = cmp_ptr(addr, (*treap)->addr);
-        if (c == 0) {
-            treap_delete_node(treap);
-            return 1;
+        else {
+            node = node->right;
         }
-        else if (c < 0) {
-            treap = &L(*treap);
+        dealloc(del);
+        _nodes--;
+        if (alpha_hi * _nodes <= alpha_lo * _maxnodes) {
+            if (_root)
+                rebalance(_root, _nodes);
+            _maxnodes = _nodes;
         }
+    }
+    void remove_aux(Node *& node, T & item)
+    {
+        if (!node)
+            return;
+        int c = Cmp::cmp(item, node->item);
+        if (c < 0)
+            remove_aux(node->left, item);
+        else if (c > 0)
+            remove_aux(node->right, item);
         else {
-            treap = &R(*treap);
+            remove_node(node);
         }
     }
-    return 0;
-}
 
-static uint64_t xorshift_rng_state = 1;
+  public:
+    BalancedTree()
+    {
+        init();
+    }
+    static BalancedTree * make()
+    {
+        BalancedTree<T, Cmp> * result = alloc<BalancedTree<T, Cmp> >();
+        result->init();
+        return result;
+    }
+    static void destroy(BalancedTree<T, Cmp> * tree)
+    {
+        tree->deinit();
+        dealloc(tree);
+    }
+    ~BalancedTree()
+    {
+        remove_all();
+    }
+    void init()
+    {
+        _nodes = _maxnodes = 0;
+        _root = NULL;
+        if (!height_to_size_init) {
+            height_to_size_init = 1;
+            double w = 1.0;
+            for (int d = 0; d < MaxTreeDepth; d++) {
+                w *= (double)alpha_hi;
+                w /= (double)alpha_lo;
+                height_to_size[d] = (Int)w;
+            }
+        }
+    }
+    void deinit()
+    {
+        remove_all();
+    }
+    void remove_all()
+    {
+        delete_nodes(_root);
+    }
+    void insert(T item)
+    {
+        insert_aux(_root, item, 0);
+        if (_nodes > _maxnodes)
+            _maxnodes = _nodes;
+    }
+    T * find(T item)
+    {
+        return find_aux(_root, item);
+    }
+    void remove(T item)
+    {
+        remove_aux(_root, item);
+    }
+    Int depth_aux(Node * node)
+    {
+        if (node == NULL)
+            return 0;
+        Int m1 = depth_aux(node->left);
+        Int m2 = depth_aux(node->right);
+        return (m1 < m2 ? m2 : m1) + 1;
+    }
+    Int depth()
+    {
+        return depth_aux(_root);
+    }
+    Int count()
+    {
+        return _nodes;
+    }
+};
+
+struct CmpAddr {
+    static int cmp(void *a, void *b) {
+        return cmp_ptr(a, b);
+    }
+};
+
+
+} // namespace JuliaGC
+
+using namespace JuliaGC;
```

Comment:
Why do we need this local namespace if we then immediately do a `using` on it? This should be explained by a comment here.

Reply:
Basically, to not pollute the global linker namespace with generic names such as `BalancedTree` or `ArrayList`. I'll add a comment.